File: DSA/Cpp/test/dsa/lib/arrays/array_test.cpp
#include <dsa/lib/arrays/array.h>
#include <dsa/lib/lists/linkedlists/ls.h>
#include <dsa/lib/trees/bt.h>
#include "array_test.h"

namespace dsa::lib::arrays {
namespace {

class ToSinglyLinkedListTest : public ArrayTest {
 protected:
    static void run(int *ary, length_t length) {
        auto head = arrayToSinglyLinkedlist(ary, length);
        EXPECT_ARRAY_EQ_LINKEDLIST(ary, length, head);
        lists::linkedlists::LinkedList::destroy(head);
    }

    static void expectNull(int *ary, length_t length) {
        auto head = arrayToSinglyLinkedlist(ary, length);
        EXPECT_EQ(nullptr, head);
    }
};
TEST_F(ToSinglyLinkedListTest, emptyArray) {
    int ary[] = {};
    expectNull(ary, arylen(ary));
}
TEST_F(ToSinglyLinkedListTest, firstElement0) {
    int ary[] = {0};
    expectNull(ary, arylen(ary));
}
TEST_F(ToSinglyLinkedListTest, firstElement0_1) {
    int ary[] = {0, 1};
    expectNull(ary, arylen(ary));
}
TEST_F(ToSinglyLinkedListTest, case1) {
    int ary[] = {1};
    run(ary, arylen(ary));
}
TEST_F(ToSinglyLinkedListTest, case2) {
    int ary[] = {1, 2, 3};
    run(ary, arylen(ary));
}

class ToCyclicSinglyLinkedListTest : public ArrayTest {
 protected:
    static void run(int *ary, length_t length, int pos, int expected, int idx) {
        ListNode *actual = toCyclicSingyLinkedList(ary, length, pos);
        EXPECT_EQ(expected, getLinkedListNodeByIndex(actual, idx)->val);
        lists::linkedlists::LinkedList::destroyCycle(actual, length);
    }
    static ListNode *toCyclicSingyLinkedList(int *ary, length_t length, int pos) {
        Array array(ary, length);
        return array.toCyclicSinglyLinkedList(pos);
    }
    static void runPosMinus(int *ary, length_t length, int pos) {
        ListNode *actual = toCyclicSingyLinkedList(ary, length, pos);
        EXPECT_ARRAY_EQ_LINKEDLIST(ary, length, actual);
        lists::linkedlists::LinkedList::destroyCycle(actual, length);
    }
};
TEST_F(ToCyclicSinglyLinkedListTest, posMinus) {
    int ary[] = {1, 2, 3};
    runPosMinus(ary, arylen(ary), -1);
}
TEST_F(ToCyclicSinglyLinkedListTest, posMinus1) {
    int ary[] = {1, 2, 3};
    runPosMinus(ary, arylen(ary), -2);
}
TEST_F(ToCyclicSinglyLinkedListTest, case1) {
    int ary[] = {1, 2, 3};
    run(ary, arylen(ary), 0, 1, 3);
}
TEST_F(ToCyclicSinglyLinkedListTest, case2) {
    int ary[] = {1, 2, 3};
    run(ary, arylen(ary), 1, 2, 3);
}
TEST_F(ToCyclicSinglyLinkedListTest, case3) {
    int ary[] = {1, 2, 3};
    run(ary, arylen(ary), 2, 3, 2);
}

class ToBinaryTreeTest : public ArrayTest {
 protected:
    static void run(int *ary, length_t length) {
        BinaryTreeNode *root = arrayToBinaryTree(ary, length);
        EXPECT_EQ_TREE(ary, ary + length, root);
        trees::BinaryTree::destroy(root);
    }
    static void expectNull(int *ary, length_t length) {
        EXPECT_EQ(nullptr, arrayToBinaryTree(ary, length));
    }
};
TEST_F(ToBinaryTreeTest, emptyArray) {
    int ary[] = {};
    expectNull(ary, arylen(ary));
}
TEST_F(ToBinaryTreeTest, firstElement0) {
    int ary[] = {0};
    expectNull(ary, arylen(ary));
}
TEST_F(ToBinaryTreeTest, firstElement0_1) {
    int ary[] = {0, 1};
    expectNull(ary, arylen(ary));
}
TEST_F(ToBinaryTreeTest, case1) {
    int ary[] = {1, 2, 3};
    run(ary, arylen(ary));
}
TEST_F(ToBinaryTreeTest, case2) {
    int ary[] = {3, 1, 5, 0, 2};
    run(ary, arylen(ary));
}
TEST_F(ToBinaryTreeTest, case3) {
    int ary[] = {5, 1, 4, 0, 6, 3, 6};
    run(ary, arylen(ary));
}
TEST_F(ToBinaryTreeTest, case4) {
    int ary[] = {1, 0};
    run(ary, arylen(ary));
}

}  // namespace
}  // namespace dsa::lib::arrays

File: DSA/Java/test/dsa/lib/arrays/ArrayTest.java
package dsa.lib.arrays;

import dsa.nodes.ListNode;
import org.junit.Test;

import static dsa.testing.DSAAssert.assertArrayEqualsLinkedList;
import static dsa.testing.DSAAssert.assertArrayEqualsTree;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class ArrayTest extends ArrayTestFx {

    public static class toSinglyLinkedListTest {
        @Test
        public void emptyArray() {
            assertNull(toLinkedList(new int[]{}));
        }

        @Test
        public void firstElement0() {
            assertNull(toLinkedList(new int[]{0}));
            assertNull(toLinkedList(new int[]{0, 1}));
        }

        @Test
        public void case1() {
            int[] ary = {1, 2, 3, 4};
            assertArrayEqualsLinkedList(ary, toLinkedList(ary));
        }

        private ListNode toLinkedList(int[] ary) {
            return arrayToSinglyLinkedlist(ary);
        }
    }

    public static class toCyclicSinglyLinkedListTest {

        @Test
        public void posMinus() {
            int[] ary = {1, 2, 3};
            assertArrayEqualsLinkedList(ary, getCyclicSinglyLinkedList(ary, -1));
        }

        @Test
        public void posMinus1() {
            int[] ary = {1, 2, 3};
            assertArrayEqualsLinkedList(ary, getCyclicSinglyLinkedList(ary, -2));
        }

        @Test
        public void case1() {
            ListNode head = getCyclicSinglyLinkedList(new int[]{1, 2}, 0);
            assertEquals(2, getLinkedListNodeByIndex(head, 3).val);
        }

        @Test
        public void case2() {
            ListNode head = getCyclicSinglyLinkedList(new int[]{1, 2, 3}, 1);
            assertEquals(2, getLinkedListNodeByIndex(head, 3).val);
        }

        @Test
        public void case3() {
            ListNode head = getCyclicSinglyLinkedList(new int[]{1, 2, 3}, 2);
            assertEquals(3, getLinkedListNodeByIndex(head, 4).val);
        }

        private ListNode getCyclicSinglyLinkedList(int[] ary, int pos) {
            return arrayToCyclicSinglyLinkedlist(ary, pos);
        }
    }

    public static class toBinaryTreeTest {
        @Test
        public void emptyArray() {
            assertNull(arrayToBinaryTree(new int[]{}));
        }

        @Test
        public void firstElement0() {
            assertNull(arrayToBinaryTree(new int[]{0}));
            assertNull(arrayToBinaryTree(new int[]{0, 1}));
        }

        @Test
        public void case1() {
            int[] array = {3, 1, 5, 0, 2};
            assertArrayEqualsTree(array, arrayToBinaryTree(array));
        }

        @Test
        public void case2() {
            int[] array = {5, 1, 4, 0, 6, 3, 6};
            assertArrayEqualsTree(array, arrayToBinaryTree(array));
        }

        @Test
        public void case3() {
            int[] ary = {1, 0};
            assertArrayEqualsTree(ary, arrayToBinaryTree(ary));
        }

    }
}
